File: Python/AoC2025/days/day8.py
import math

class Day8:
    def __init__(self):
        with open("../../Data/2025/day8.txt") as f:
            self.data = [line.rstrip('\n') for line in f]
        self.positions = [[int(x) for x in line.split(',')] for line in self.data]
        self.distances = self.calculate_distances()

    def part_one(self):
        lengths = [len(circuit) for circuit in self.create_circuits()]
        lengths.sort()
        return lengths[-1] * lengths[-2] * lengths[-3]
    
    def part_two(self):
        start, end = self.create_full_circuit()
        x1, x2 = self.positions[start][0], self.positions[end][0]
        return x1 * x2
    
    def euclidean_distance(self, x1, y1, z1, x2, y2, z2):
        return math.sqrt((x1 - x2) ** 2 + (y1 - y2) ** 2 + (z1 - z2) ** 2)
    
    def calculate_distances(self):
        distances = []
        for i in range(len(self.positions) - 1):
            x1, y1, z1 = self.positions[i][0], self.positions[i][1], self.positions[i][2]
            for j in range(i + 1, len(self.positions)):
                x2, y2, z2 = self.positions[j][0], self.positions[j][1], self.positions[j][2]
                distances.append([self.euclidean_distance(x1, y1, z1, x2, y2, z2), i, j])
        distances.sort()
        return distances
    
    def create_circuits(self):
        circuits = [[self.distances[0][1], self.distances[0][2]]]
        for distance in self.distances[1:1000]:
            start, end = distance[1], distance[2]
            connected = {start, end}
            not_connected = []
            for circuit in circuits:
                if (start in circuit) or (end in circuit):
                    connected.update(circuit)
                else:
                    not_connected.append(circuit)
            not_connected.append(list(connected))
            circuits = not_connected
        return circuits
    
    def create_full_circuit(self):
        circuits = [[self.distances[0][1], self.distances[0][2]]]
        for distance in self.distances[1:]:
            start, end = distance[1], distance[2]
            connected = {start, end}
            not_connected = []
            for circuit in circuits:
                if (start in circuit) or (end in circuit):
                    connected.update(circuit)
                else:
                    not_connected.append(circuit)
            if len(not_connected) == 0:
                return start, end
            not_connected.append(list(connected))
            circuits = not_connected
        return -1, -1

    def print_solution(self):
        print("Day 8:")
        print(f"Part 1: {self.part_one()}")
        print(f"Part 2: {self.part_two()}")

File: CSharp/AoC2025/Days/Day8.cs
namespace AoC2025;

public static class Day8
{

    public static int PartOne()
    {
        var lengths = CreateCircuits(1000, false, out int start, out int end)
            .Select(h => h.Count)
            .ToArray();
        lengths.Sort((a, b) => b - a);
        return lengths[0] * lengths[1] * lengths[2];
    }

    public static int PartTwo()
    {
        var _ = CreateCircuits(-1, true, out int start, out int end);
        var data = GetData().ToArray();
        return data[start].First() * data[end].First();
    }
    
    // Helper methods

    private static List<HashSet<int>> CreateCircuits(int length, bool isComplete, out int start, out int end)
    {
        var distances = GetOrderedDistances();
        if (length == -1)
            length = distances.Count;
        List<HashSet<int>> circuits = [[distances[0][1], distances[0][2]]];
        foreach (var distance in distances[1..length])
        {
            start = distance[1];
            end = distance[2];
            List<HashSet<int>> carryOverCircuits = [];
            HashSet<int> mergedCircuits = [start, end];

            foreach (var circuit in circuits)
            {
                if (circuit.Contains(start) || circuit.Contains(end))
                {
                    mergedCircuits.UnionWith(circuit);
                }
                else
                {
                    carryOverCircuits.Add(circuit);
                }
            }

            if (isComplete && carryOverCircuits.Count == 0)
                return [];

            carryOverCircuits.Add(mergedCircuits);
            circuits = carryOverCircuits;
        }
        start = end = -1;
        return circuits;
    }

    // Data methods

    private static IEnumerable<IEnumerable<int>> GetData() => File.ReadAllLines("../../Data/2025/day8.txt")
        .Select(s => s.Split(',').Select(c => int.Parse(c)));

    private static int EuclideanDistance(IEnumerable<int> a, IEnumerable<int> b)
    {
        if (a.Count() != 3 || b.Count() != 3)
            throw new ArgumentException("Coordinates must have exactly x, y, z.");

        long total = Enumerable.Range(0, 3)
            .Select(num => (long)Math.Pow((a.Skip(num).First() - b.Skip(num).First()), 2))
            .Sum();
        return (int)Math.Sqrt(total);
    }

    private static List<int[]> GetOrderedDistances()
    {
        var data = GetData().ToArray();
        List<int[]> distances = [];

        for (int i = 0; i < data.Length - 1; i++)
        {
            for (int j = i + 1; j < data.Length; j++)
            {
                int d = EuclideanDistance(data[i], data[j]);
                int[] distance = [d, i, j];
                distances.Add(distance);
            }
        }

        distances.Sort((a, b) => a[0] - b[0]);
        return distances;
    }
}